cpp/convert_json_to_marc.cc

```diff
@@ -0,0 +1,493 @@
+                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_prefixes,
+                                 const std::vector<std::pair<char, std::string>> &subfield_codes_to_fixed_subfields,
+                                 const std::string &json_tag, const std::string &field_contents_prefix, const bool required)
+    : name_(name), tag_(tag), overflow_tag_(overflow_tag), indicator1_(indicator1), indicator2_(indicator2), repeat_field_(repeat_field),
+      subfield_codes_to_json_tags_(subfield_codes_to_json_tags), subfield_codes_to_prefixes_(subfield_codes_to_prefixes),
+      subfield_codes_to_fixed_subfields_(subfield_codes_to_fixed_subfields), json_tag_(json_tag),
+      field_contents_prefix_(field_contents_prefix), required_(required)
+{
+    if (not overflow_tag_.empty() and repeat_field_)
+        LOG_ERROR("field \"" + name_ + "\" can't have both, an over flow tag and being a repeat field!");
+
+    if (subfield_codes_to_json_tags_.empty() and json_tag_.empty())
+        LOG_ERROR("field \"" + name_
+                  + "\" neither a mapping to the contents of a control field nor the contents of data subfields has been specified!");
+}
+
+
+class JSONNodeToBibliographicLevelMapper {
+    std::string json_tag_;
+    MARC::Record::BibliographicLevel default_;
+    std::vector<std::pair<RegexMatcher *, MARC::Record::BibliographicLevel>> regex_to_bibliographic_level_map;
+public:
+    JSONNodeToBibliographicLevelMapper(const std::string &item_type_tag, const std::string &item_type_map);
+    MARC::Record::BibliographicLevel getBibliographicLevel(const JSON::ObjectNode &object_node) const;
+    MARC::Record::BibliographicLevel getBibliographicLevel(const std::string &string_value) const;
+};
+
+
+// Expects two values separated by a colon.  Backslash-escaped colons do not count as the separator colon!
+void SplitPatternAndItemType(const std::string &pattern_and_type, std::string * const pattern, std::string * const type) {
+    bool escaped(false), in_pattern(true);
+    for (const char ch : pattern_and_type) {
+        if (escaped) {
+            if (in_pattern)
+                *pattern += ch;
+            else
+                *type += ch;
+            escaped = false;
+        } else if (ch == '\\')
+            escaped = true;
+        else if (ch == ':') {
+            if (not in_pattern)
+                LOG_ERROR("bad regex pattern to item type mapping with additional unescaped colon: \"" + pattern_and_type + "\"!");
+            in_pattern = false;
+        } else {
+            if (in_pattern)
+                *pattern += ch;
+            else
+                *type += ch;
+            escaped = false;
+        }
+    }
+
+    if (escaped)
+        LOG_ERROR("bad regex pattern to item type mapping: \"" + pattern_and_type + "\"!");
+}
+
+
+MARC::Record::BibliographicLevel MapTypeStringToBibliographicLevel(const std::string &item_type) {
+    if (item_type == "monograph")
```

Comment:
Since this is supposed to be a generic tool, wouldn't it be better to have the item type strings saved in the config file? A different JSON schema might use other identifiers for those bibliographic levels.

Reply:
There is only a fixed set of possible bibliographic levels as per the MARC-21 standard.  There are still various other possible problems with this tool not being fully generic.  I decided to not go much beyond what is needed for the data from CORE for now.  We may never need to support anything else and if we do it will be easier to decide what has to be extended.  Also we don't have any outside users of our tool so we don't care about how generic it is.